src/utils/generator.js

```diff
@@ -137,17 +141,33 @@ export class TemporalBuildNodeEntry extends GeneratorEntry {
     }
   }
 
-  declared: void | AbstractValue | ConcreteValue;
+  declared: void | Value;
   args: Array<Value>;
   // If we're just trying to add roots for the serializer to notice, we don't need a buildNode.
   buildNode: void | GeneratorBuildNodeFunction;
   dependencies: void | Array<Generator>;
   isPure: void | boolean;
   mutatesOnly: void | Array<Value>;
+  optimizationFn:
+    | void
+    | ((VisitEntryCallbacks, Generator, TemporalBuildNodeEntry) => TemporalBuildNodeEntryOptimizationStatus);
 
   visit(callbacks: VisitEntryCallbacks, containingGenerator: Generator): boolean {
     let omit = this.isPure && this.declared && callbacks.canOmit(this.declared);
 
+    if (this.optimizationFn !== undefined && this.declared !== undefined) {
+      let optimizationStatus = this.optimizationFn(callbacks, containingGenerator, this);
+      // If we have a possible optimizaiton, we need to delay this entry visit
+      // and re-visit later as delayed entry.
+      if (optimizationStatus === "POSSIBLE_OPTIMIZATION") {
+        callbacks.recordDelayedEntry(containingGenerator, this);
+        return false;
+      } else if (optimizationStatus === "OPTIMIZATION") {
+        // We don't want to visit this entry anymore as we've optimized and created
```

Comment:
Seems a bit arbitrary that in this case `replaceAndRecordDelayedEntry` is called by the `optimizationFn`, but in the `"POSSIBLE_OPTIMIZATION"` function the `recordDelayedEntry` call happens here.